File: dev/exact_nn/moe/moe_parity.js
// moe_parity.js — verify moeForward.js (JS deploy) reproduces the trained torch
// MoE on dumped samples (moe_parity.py). Reports max |logit_js - logit_torch| and
// argmax/argmin agreement (the only thing that affects S_dec).
//   node moe_parity.js moe_weights.json parity.json
'use strict';
const fs = require('fs');
const { loadMoE } = require('./moeForward.js');

const moe = loadMoE(process.argv[2] || 'moe_weights.json');
const samples = JSON.parse(fs.readFileSync(process.argv[3] || 'parity.json', 'utf8'));

let maxAbs = 0, maxG = 0, argOk = 0, n = 0;
const worst = [];
for (const s of samples) {
    const { logit, g } = moe.forward(s.planner_block, s.endgame_block, s.gate_feat,
        s.slot_valid.map(Boolean), s.p_valid.map(Boolean), s.e_valid.map(Boolean));
    let mx = 0;
    for (let k = 0; k < 16; k++) {
        if (!s.slot_valid[k]) continue;
        const d = Math.abs(logit[k] - s.torch_logit[k]);
        if (d > mx) mx = d;
    }
    maxAbs = Math.max(maxAbs, mx);
    maxG = Math.max(maxG, Math.abs(g - s.torch_g));
    // argmax over valid slots, both sides
    const am = (arr) => { let bi = -1, bs = -Infinity; for (let k = 0; k < 16; k++) if (s.slot_valid[k] && arr[k] > bs) { bs = arr[k]; bi = k; } return bi; };
    if (am(logit) === am(s.torch_logit)) argOk++;
    else worst.push({ regime: s.regime, js: am(logit), torch: am(s.torch_logit), maxAbs: mx });
    n++;
}
console.log(`parity: n=${n} maxAbs(logit)=${maxAbs.toExponential(3)} maxAbs(gate)=${maxG.toExponential(3)} argmaxAgree=${(argOk / n).toFixed(4)}`);
if (worst.length) console.log('argmax disagreements:', JSON.stringify(worst.slice(0, 10)));

File: dev/exact_nn/moe/moe_parity.py
#!/usr/bin/env python3
"""Dump torch MoE forward on sample decisions -> JSON, for the JS parity check
(moe_parity.js). Confirms moeForward.js reproduces the trained model bit-closely.
  python3 moe_parity.py --pack pack --ckpt moe_ckpt.pt --n 200 --out parity.json
"""
import argparse, json
import numpy as np
import torch
from moe_model import MoEPolicy, PLANNER_DIM, ENDGAME_DIM, NSLOT
from moe_train import load_pack, Pack


def main():
    ap = argparse.ArgumentParser()
    ap.add_argument('--pack', required=True)
    ap.add_argument('--ckpt', required=True)
    ap.add_argument('--n', type=int, default=200)
    ap.add_argument('--out', default='parity.json')
    args = ap.parse_args()
    ck = torch.load(args.ckpt, map_location='cpu')
    model = MoEPolicy(d=ck['args']['d'])
    model.load_state_dict(ck['model']); model.eval()
    d, pm, em = load_pack(args.pack)
    pk = Pack(d, torch.device('cpu'))
    g = torch.Generator().manual_seed(123)

    samples = []
    # planner samples
    pidx = pk.p_va[torch.randint(0, len(pk.p_va), (args.n,), generator=g).numpy()]
    eidx = pk.e_va[torch.randint(0, len(pk.e_va), (args.n,), generator=g).numpy()]
    for which, idx in (('planner', pidx), ('endgame', eidx)):
        for j in idx[:args.n]:
            if which == 'planner':
                b = pk.assemble(torch.tensor([j]), torch.empty(0, dtype=torch.long))
            else:
                b = pk.assemble(torch.empty(0, dtype=torch.long), torch.tensor([j]))
            with torch.no_grad():
                logit, gg = model(b['pb'], b['eb'], b['gf'], b['sv'], b['pv'], b['ev'])
            samples.append(dict(
                regime=which,
                planner_block=b['pb'][0].tolist(), endgame_block=b['eb'][0].tolist(),
                gate_feat=b['gf'][0].tolist(), slot_valid=b['sv'][0].int().tolist(),
                p_valid=b['pv'][0].int().tolist(), e_valid=b['ev'][0].int().tolist(),
                torch_logit=logit[0].tolist(), torch_g=float(gg.reshape(-1)[0]),
            ))
    json.dump(samples, open(args.out, 'w'))
    print('wrote', args.out, len(samples), 'samples')


if __name__ == '__main__':
    main()
